src/main/java/ch/njol/skript/expressions/ExprBannerPatterns.java

```diff
@@ -0,0 +1,271 @@
+
+	static {
+		Skript.registerExpression(ExprBannerPatterns.class, Pattern.class, ExpressionType.PROPERTY,
+			"[the] banner pattern[s] of %itemstacks/itemtypes/slots/blocks%",
+			"%itemstacks/itemtypes/slots/blocks%'[s] banner pattern[s]",
+			"[the] %integer%[st|nd|rd|th] [banner] pattern of %itemstacks/itemtypes/slots/blocks%",
+			"%itemstacks/itemtypes/slots/blocks%'[s] %integer%[st|nd|rd|th] [banner] pattern"
+		);
+	}
+
+	private Expression<?> objects;
+	private Expression<Integer> patternNumber;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (matchedPattern <= 1) {
+			objects = exprs[0];
+		} else if (matchedPattern == 2) {
+			//noinspection unchecked
+			patternNumber = (Expression<Integer>) exprs[0];
+			objects = exprs[1];
+		} else {
+			//noinspection unchecked
+			patternNumber = (Expression<Integer>) exprs[1];
+			objects = exprs[0];
+		}
+		setExpr(objects);
+		return true;
+	}
+
+	@Override
+	protected Pattern @Nullable [] get(Event event, Object[] source) {
+		List<Pattern> patterns = new ArrayList<>();
+		Integer placement = patternNumber != null ? patternNumber.getSingle(event) : null;
+		for (Object object : objects.getArray(event)) {
+			if (object instanceof Block block) {
+				if (block.getState() instanceof Banner banner) {
+					if (placement != null) {
+						if (banner.numberOfPatterns() >= placement)
+							patterns.add(banner.getPattern(placement - 1));
+					} else {
+						patterns.addAll(banner.getPatterns());
+					}
+				}
+			} else {
+				ItemStack itemStack = ItemUtils.asItemStack(object);
+				if (itemStack == null || !(itemStack.getItemMeta() instanceof BannerMeta bannerMeta))
+					continue;
+				if (placement != null) {
+					if (bannerMeta.numberOfPatterns() >= placement)
+						patterns.add(bannerMeta.getPattern(placement - 1));
+				} else {
+					patterns.addAll(bannerMeta.getPatterns());
+				}
+			}
+		}
+		return patterns.toArray(new Pattern[0]);
+	}
+
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		return switch (mode) {
+			case SET, DELETE -> {
+				if (patternNumber != null)
+					yield CollectionUtils.array(Pattern.class);
+				yield CollectionUtils.array(Pattern[].class);
+			}
+			case REMOVE, ADD -> CollectionUtils.array(Pattern[].class);
+			default -> null;
+		};
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		Pattern[] patterns = null;
+		Pattern pattern = null;
+		Integer placement = null;
+		if (patternNumber != null) {
+			placement = patternNumber.getSingle(event);
+			if (delta != null && delta[0] != null) {
+				pattern = (Pattern) delta[0];
+			}
+		} else if (delta != null) {
+			patterns = (Pattern[]) delta;
+		}
+		Integer finalPlacement = placement;
+		Pattern finalPattern = pattern;
+		List<Pattern> patternList = patterns != null ? Arrays.stream(patterns).toList() : new ArrayList<>();
```

Comment:
this can be simplified
we don't need both pattern and patterns, just use a single-element array if it's 1 pattern